# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [17]:
df = pd.read_csv('Step3_dataset.csv',index_col = 'Unnamed: 0')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         1  
1               90.0             0.0         1  
2              152.0            30.0         1  
3              122.0             0.0         0  
4              104.0            80.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [18]:
df.dtypes

Name                  object
state                 object
summit_elev            int64
vertical_drop          int64
trams                  int64
fastEight            float64
fastSixes              int64
fastQuads              int64
quad                   int64
triple                 int64
double                 int64
surface                int64
total_chairs           int64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object

In [19]:
df1 = pd.concat([df,pd.get_dummies(df['state'])], axis=1).drop('state', axis=1)
df1

Name  summit_elev  vertical_drop  trams  \
0                       Alyeska Resort         3939           2500      1   
1                  Eaglecrest Ski Area         2600           1540      0   
2                     Hilltop Ski Area         2090            294      0   
3                     Arizona Snowbowl        11500           2300      0   
4                  Sunrise Park Resort        11100           1800      0   
..                                 ...          ...            ...    ...   
325               Meadowlark Ski Lodge         9500           1000      0   
326          Sleeping Giant Ski Resort         7428            810      0   
327                   Snow King Resort         7808           1571      0   
328  Snowy Range Ski & Recreation Area         9663            990      0   
329                White Pine Ski Area         9500           1100      0   

     fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0.0          0          2     2       0       0  ...             0   
1          0.0          0          0     0       0       4  ...             0   
2          0.0          0          0     0       1       0  ...             0   
3          0.0          1          0     2       2       1  ...             0   
4          0.0          0          1     2       3       1  ...             0   
..         ...        ...        ...   ...     ...     ...  ...           ...   
325        0.0          0          0     0       1       1  ...             0   
326        0.0          0          0     0       1       1  ...             0   
327        0.0          0          0     1       1       1  ...             0   
328        0.0          0          0     0       1       3  ...             0   
329        0.0          0          0     0       2       0  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
325             0          0     0        0         0           0   
326             0          0     0        0         0           0   
327             0          0     0        0         0           0   
328             0          0     0        0         0           0   
329             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
325              0          0        1  
326              0          0        1  
327              0          0        1  
328              0          0        1  
329              0          0        1  

[330 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [20]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [21]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25, random_state=1)

In [22]:
print(X.shape, y.shape)

(330, 58) (330,)


Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [42]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error, mean_squared_error, r2_score
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [43]:
#Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [44]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
print(explained_variance_score(y_test,y_pred),mean_absolute_error(y_test,y_pred),r2_score(y_test,y_pred))

0.9427818561042524 4.823344177064809 0.9425369820827088


In [26]:
from math import sqrt
rmse = sqrt(mean_squared_error(y_test, y_pred))
rmse

6.262254188252109

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [27]:
lm.intercept_

14.513524776512412

In [40]:
import statsmodels.api as sm
x = sm.add_constant(X_train)
results = sm.OLS(y_train,x).fit()
results.summary()

C:\Users\SiddharthaMisra\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.912
Model:                            OLS   Adj. R-squared:                  0.886
Method:                 Least Squares   F-statistic:                     35.06
Date:                Wed, 03 Jun 2020   Prob (F-statistic):           5.06e-75
Time:                        12:09:29   Log-Likelihood:                -804.79
No. Observations:                 247   AIC:                             1724.
Df Residuals:                     190   BIC:                             1924.
Df Model:                          56                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                14.1104      3.853      3.662      0.000       6.510      21.710
summit_elev       -3.441e-06      0.001     -0.004      0.997      -0.002       0.002
vertical_drop         0.0016      0.002      1.016      0.311      -0.001       0.005
trams                -0.8745      1.660     -0.527      0.599      -4.149       2.400
fastEight             2.7954      8.392      0.333      0.739     -13.758      19.348
fastSixes            -0.3534      1.521     -0.232      0.817      -3.354       2.647
fastQuads            -0.3149      1.102     -0.286      0.775      -2.489       1.860
quad                  0.2370      1.179      0.201      0.841      -2.088       2.562
triple                0.4967      1.176      0.422      0.673      -1.823       2.816
double               -0.1551      1.129     -0.137      0.891      -2.381       2.071
surface              -1.1097      1.253     -0.886      0.377      -3.581       1.362
total_chairs          0.7215      1.129      0.639      0.524      -1.506       2.949
Runs                 -0.0109      0.037     -0.295      0.768      -0.084       0.062
TerrainParks         -0.0860      0.327     -0.263      0.793      -0.731       0.559
LongestRun_mi        -0.2958      0.720     -0.411      0.682      -1.715       1.124
SkiableTerrain_ac    -0.0017      0.001     -1.133      0.259      -0.005       0.001
Snow Making_ac        0.0025      0.005      0.472      0.637      -0.008       0.013
daysOpenLastYear      0.0235      0.033      0.713      0.477      -0.041       0.088
yearsOpen             0.0034      0.004      0.899      0.370      -0.004       0.011
averageSnowfall       0.0011      0.008      0.131      0.896      -0.015       0.017
AdultWeekday          0.8057      0.031     26.359      0.000       0.745       0.866
projectedDaysOpen    -0.0115      0.035     -0.332      0.741      -0.080       0.057
NightSkiing_ac        0.0087      0.006      1.346      0.180      -0.004       0.021
clusters             -1.4472      1.307     -1.108      0.269      -4.025       1.130
Alaska               -3.1006      5.284     -0.587      0.558     -13.524       7.323
Arizona              -4.3277      8.567     -0.505      0.614     -21.226      12.571
California           -3.5156      4.128     -0.852      0.395     -11.658       4.627
Colorado             -2.7556      4.931     -0.559      0.577     -12.482       6.971
Connecticut           3.3889      4.305      0.787      0.432      -5.104      11.881
Idaho                -4.5962      3.665     -1.254      0.211     -11.826       2.633
Illinois             -0.1188      4.830     -0.025      0.980      -9.647       9.409
Indiana              -6.2136      5.546     -1.120      0.264     -17.154       4.726
Iowa                 -4.4074      5.412     -0.814      0.416     -15.082

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [28]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
Tennessee            20.595079
North Carolina       14.010776
West Virginia        10.070417
Maryland              9.500111
Ohio                  7.292454
Montana               7.090454
Massachusetts         6.694864
New Mexico            6.668592
Indiana               6.616756
Virginia              6.535344
Idaho                 4.999354
South Dakota          4.818446
Iowa                  4.810537
Arizona               4.730885
New Hampshire         4.613200
Minnesota             4.234379
Missouri              3.929704
California            3.918708
Oregon                3.658979
Alaska                3.503785
Wyoming               3.489753
Colorado              3.158780
Connecticut           2.985732
fastEight             2.795441
Rhode Island          2.276764
Washington            2.073749
Pennsylvania          1.940286
Michigan              1.874696
clusters              1.447205
New York              1.287509
surface               1.109664
trams                 0.874478
AdultWeekday          0.805689
total_chairs          0.721536
Illinois              0.521956
Maine                 0.521730
triple                0.496713
Vermont               0.463963
fastSixes             0.353436
Wisconsin             0.317521
fastQuads             0.314949
Nevada                0.304070
LongestRun_mi         0.295813
quad                  0.236965
Utah                  0.189158
New Jersey            0.168432
double                0.155056
TerrainParks          0.085999
daysOpenLastYear      0.023484
projectedDaysOpen     0.011548
Runs                  0.010878
NightSkiing_ac        0.008667
yearsOpen             0.003426
Snow Making_ac        0.002541
SkiableTerrain_ac     0.001697
vertical_drop         0.001574
averageSnowfall       0.001051
summit_elev           0.000003

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [47]:
X2 = df.drop(['Name','state','AdultWeekend'], axis=1)
X2_train, X2_test, y2_train, y2_test = train_test_split(X2,y, test_size=0.25, random_state=1)
lm2 = linear_model.LinearRegression()
model = lm2.fit(X2_train,y2_train)

In [48]:
y2_pred = model.predict(X2_test)
print(explained_variance_score(y2_test,y2_pred),mean_absolute_error(y2_test,y2_pred),r2_score(y2_test,y2_pred))

0.9241713964368776 5.639050353390269 0.9236300986148854


In [49]:
mean_absolute_error(y2_test,y2_pred)

5.639050353390269

In [50]:
pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
fastEight             2.606555
surface               1.360366
clusters              1.327564
trams                 0.931461
total_chairs          0.874149
AdultWeekday          0.826816
fastSixes             0.703678
double                0.495427
quad                  0.325099
triple                0.203371
fastQuads             0.177300
TerrainParks          0.154340
LongestRun_mi         0.136623
daysOpenLastYear      0.041332
projectedDaysOpen     0.027042
NightSkiing_ac        0.009093
averageSnowfall       0.007177
SkiableTerrain_ac     0.003224
yearsOpen             0.002712
Runs                  0.002211
vertical_drop         0.002170
Snow Making_ac        0.001414
summit_elev           0.000247

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [56]:
X3 = X2.drop(['summit_elev'], axis=1)
X3_train, X3_test, y3_train, y3_test = train_test_split(X3,y, test_size=0.25, random_state=1)
lm3 = linear_model.LinearRegression()
model = lm3.fit(X3_train,y3_train)

In [57]:
y3_pred = model.predict(X3_test)
print(explained_variance_score(y3_test,y3_pred),mean_absolute_error(y3_test,y3_pred),r2_score(y3_test,y3_pred))

0.9195591436647751 5.7994627061555395 0.9189849794630058


In [58]:
mean_absolute_error(y3_test,y3_pred)

5.7994627061555395

In [59]:
pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
fastEight             3.392771
surface               1.495389
clusters              1.215330
total_chairs          1.003425
trams                 0.921393
AdultWeekday          0.824754
double                0.612396
fastSixes             0.561954
quad                  0.251930
fastQuads             0.247959
TerrainParks          0.142745
LongestRun_mi         0.129629
triple                0.073907
daysOpenLastYear      0.037390
projectedDaysOpen     0.022270
NightSkiing_ac        0.010102
averageSnowfall       0.009704
SkiableTerrain_ac     0.003416
yearsOpen             0.002923
vertical_drop         0.001707
Runs                  0.001633
Snow Making_ac        0.001589

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.943 | 4.823 |'base_elev'|
| Model 2. | 0.924 | 5.639 |'state','base_elev'|
| Model 3. | 0.920 | 5.800 |'state','summit_elev','base_elev'|

Model Selection:

Since we cannot move Big Mountain Resort to a different mountain or location, Model 3 is the logical choice for a model. It might be helpful for the BMR stakeholders to see the variance in lift-ticket prices across states, but that question is not very relevant to the chair-lift issue at hand. Therefore, we could have eliminated base elevation, summit elevation, and state during the EDA phase, as our question was how to modify BMR's practices, not to examine industry-wide trends in pricing.